### ISSUE TITLE
Incorrect EOS masking

### ISSUE THREAD

**Issue by teoremma** [epfl-dlab/transformers-CFG#105]
Probably related to #103 

Replicate with
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers_cfg.grammar_utils import IncrementalGrammarConstraint
from transformers_cfg.generation.logits_process import GrammarConstrainedLogitsProcessor

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model_id = "mistralai/Mistral-7B-Instruct-v0.3"

tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto")
model.eval()

grammar_str = 'root ::= "" | "0" root | "1" root'
grammar_constraint = IncrementalGrammarConstraint(grammar_str, "root", tokenizer)
logits_processor = GrammarConstrainedLogitsProcessor(grammar_constraint)

prompt = "Generate an arbitrary binary string."
messages = [{"role": "user", "content": prompt}]
input_ids = tokenizer.apply_chat_template(messages, add_generation_prompt=True, return_tensors="pt").to(model.device)

output = model.generate(
    input_ids,
    max_new_tokens=4,
    eos_token_id=tokenizer.eos_token_id,
    pad_token_id=tokenizer.pad_token_id,
    do_sample=True,
    logits_processor=[logits_processor],
    return_dict_in_generate=True,
    output_scores=True,
)

output_ids = output.sequences[:, input_ids.shape[-1] :]
output_str = tokenizer.decode(output_ids[0], skip_special_tokens=True)
print(f"Output: {output_str}")
print()

output_scores = torch.stack(output.scores).transpose(0, 1)
for i in range(output_ids.shape[1]):
    top_scores, top_indices = torch.topk(output_scores[0][i], k=5)
    print(f"Step {i}")
    for score, id in zip(top_scores, top_indices):
        print(f"{repr(tokenizer.decode(id))} ({id}): {score.item()}")
    print()
```

Output is
```
Output: 1010

Step 0
'1' (29508): 1.3009755611419678
'0' (29502): 0.7234944105148315
'</s>' (2): -15.079431533813477
'<s>' (1): -inf
'<unk>' (0): -inf

Step 1
'0' (29502): 21.700206756591797
'1' (29508): 19.28874397277832
'</s>' (2): -inf
'<s>' (1): -inf
'<unk>' (0): -inf

Step 2
'1' (29508): 24.555517196655273
'0' (29502): 17.726497650146484
'</s>' (2): -inf
'<s>' (1): -inf
'<unk>' (0): -inf

Step 3
'0' (29502): 21.117769241333008
'1' (29508): 19.760522842407227
'</s>' (2): -inf
'<s>' (1): -inf
'<unk>' (0): -inf
```

EOS is being masked at every step after the first, although it should be allowed always.

**Comment by Saibo-creator** [comment]
Thanks for the detailed report! I can reproduce the issue and will look into fixing it.

**Comment by Saibo-creator** [comment]
Hi @teoremma 
Could you let me know which version of transformers-cfg you’re using or which commit you’re on?

When I run your script, I actually get a different output:
```
Output: 1010

Step 0
'1' (29508): 1.3009755611419678
'0' (29502): 0.7234944105148315
'</s>' (2): -15.079431533813477
'<s>' (1): -inf
'<unk>' (0): -inf

Step 1
'0' (29502): 21.700206756591797
'</s>' (2): -inf
'<unk>' (0): -inf
'[INST]' (3): -inf
'<s>' (1): -inf

Step 2
'1' (29508): 24.555517196655273
'</s>' (2): -inf
'<unk>' (0): -inf
'[INST]' (3): -inf
'<s>' (1): -inf

Step 3
'0' (29502): 21.117769241333008
'</s>' (2): -inf
'<unk>' (0): -inf
'[INST]' (3): -inf
'<s>' (1): -inf
```
In my output, it seems like 0 and 1 should both be allowed, but only one appears at each step, which is quite problematic :') 
It looks like your version is actually working better, at least it allows both 1 and 0 .

Would you mind sharing your version or commit hash? That would help a lot!

Thanks! 😊

**Comment by Saibo-creator** [comment]
This issue is fixed in the PR #108 , now the output is 


```
Output: 1010

Step 0
'1' (29508): 1.3009755611419678
'0' (29502): 0.7234944105148315
'</s>' (2): -15.079431533813477
'<s>' (1): -inf
'<unk>' (0): -inf

Step 1
'0' (29502): 21.700206756591797
'1' (29508): 19.28874397277832
'</s>' (2): -12.689655303955078
'<s>' (1): -inf
'<unk>' (0): -inf

Step 2
'1' (29508): 24.555517196655273
'0' (29502): 17.726497650146484
'</s>' (2): -9.426843643188477
'<s>' (1): -inf
'<unk>' (0): -inf

Step 3
'0' (29502): 21.117769241333008
'1' (29508): 19.760522842407227
'</s>' (2): -9.650508880615234
'<s>' (1): -inf
'<unk>' (0): -inf
```

You can install from the main branch `pip install git+https://github.com/epfl-dlab/transformers-CFG.git@main` to test. 

We will include this in the next release, which will come probably next week after some other fixes :)
